File: CS/WindowsApplication1/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors.Repository;

namespace WindowsApplication1
{
    public partial class Form1 : Form
    {
        Type[] types = new Type[] { typeof(string), typeof(int), typeof(bool), typeof(DateTime) };
        DataTable tbl;
                private DataTable CreateTable(int RowCount)
        {
            tbl = new DataTable();
            tbl.Columns.Add("Object", typeof(object));
            for (int i = 0; i < RowCount; i++)
                tbl.Rows.Add(new object[] { null });
            return tbl;
        }
        

        public Form1()
        {
            InitializeComponent();
            gridControl1.DataSource = CreateTable(20);
            dataGridView1.DataSource = gridControl1.DataSource;
        }

        Type GetDataTypeByIndex(int rowHandle)
        {
            return types[rowHandle % types.Length];
        }

        RepositoryItem GetRIByIndex(int rowHandle)
        {
            switch (rowHandle % types.Length)
            {
                case 0: return repositoryItemTextEdit1;
                case 1: return repositoryItemSpinEdit1;
                case 2: return repositoryItemCheckEdit1;
                case 3: return repositoryItemDateEdit1;
                default: return repositoryItemTextEdit1;
            }
        }

        private void gridView1_CustomUnboundColumnData(object sender, DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs e)
        {
            if (e.IsGetData) e.Value = GetDataTypeByIndex(e.ListSourceRowIndex);
        }

        private void gridView1_CustomRowCellEdit(object sender, DevExpress.XtraGrid.Views.Grid.CustomRowCellEditEventArgs e)
        {
            if (e.Column != colDataType)
                e.RepositoryItem = GetRIByIndex(e.RowHandle);
        }
    }
}

File: VB/WindowsApplication1/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraEditors.Repository

Namespace WindowsApplication1
	Partial Public Class Form1
		Inherits Form
		Private types() As Type = { GetType(String), GetType(Integer), GetType(Boolean), GetType(DateTime) }
		Private tbl As DataTable
				Private Function CreateTable(ByVal RowCount As Integer) As DataTable
			tbl = New DataTable()
			tbl.Columns.Add("Object", GetType(Object))
			For i As Integer = 0 To RowCount - 1
				tbl.Rows.Add(New Object() { Nothing })
			Next i
			Return tbl
				End Function


		Public Sub New()
			InitializeComponent()
			gridControl1.DataSource = CreateTable(20)
			dataGridView1.DataSource = gridControl1.DataSource
		End Sub

		Private Function GetDataTypeByIndex(ByVal rowHandle As Integer) As Type
			Return types(rowHandle Mod types.Length)
		End Function

		Private Function GetRIByIndex(ByVal rowHandle As Integer) As RepositoryItem
			Select Case rowHandle Mod types.Length
				Case 0
					Return repositoryItemTextEdit1
				Case 1
					Return repositoryItemSpinEdit1
				Case 2
					Return repositoryItemCheckEdit1
				Case 3
					Return repositoryItemDateEdit1
				Case Else
					Return repositoryItemTextEdit1
			End Select
		End Function

		Private Sub gridView1_CustomUnboundColumnData(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs) Handles gridView1.CustomUnboundColumnData
			If e.IsGetData Then
				e.Value = GetDataTypeByIndex(e.ListSourceRowIndex)
			End If
		End Sub

		Private Sub gridView1_CustomRowCellEdit(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Grid.CustomRowCellEditEventArgs) Handles gridView1.CustomRowCellEdit
			If e.Column IsNot colDataType Then
				e.RepositoryItem = GetRIByIndex(e.RowHandle)
			End If
		End Sub
	End Class
End Namespace